grep-0018-modernize-c++.md

```diff
@@ -0,0 +1,417 @@
+# GREP 0018 -- Modernize C++
+
+- Original Author: Thomas Habets <thomas@habets.se>
+- Champion: Thomas Habets <thomas@habets.se>
+- Status: Draft
+
+History:
+- 20-Jan-2020: Initial Draft
+
+## Abstract
+
+As of early 2020 there is lots of pre-C++11 code in GNU Radio, and
+some other pattern of non-idiomatic C++.
+
+This GREP lays out how to get from here to there with minimal
+breaking.
+
+## Copyright / License
+
+This GREP is licensed as CC-BY-ND. Copyright 2020 Google Inc.
+
+## Motivation
+
+Cleaning up whole patterns of the code 
+
+The legacy patterns this GREP lays out a plan for making the code base:
+
+* more readable: standard components, less code, and explicit const
+* safer: less manual memory management, prevent inherently broken copy
+* faster: fewer pointer indirections, more information to the compiler
+
+## Description
+
+### Open questions
+
+* What level of ABI guarantee does GNU Radio make between and within
+  minor releases?
+
+### More const
+
+Benefits:
+* more readable; even in a function reader can see that a name has a
+  final value
+* avoids accidents
+* possibly allows compiler optimizations
+
+Recommendation:
+* private members and other `_impl.xx` code should be free to change
+  at any time to add const
+* adding const to exposed interfaces can be done between minor
+  releases
+
+Risks: none.
+
+### More constexpr
```

Comment:
What is the action recommended here? Make `constexpr` recommendations part of our coding guidelines, or actually go and change stuff?

Reply:
Clarifying…